assembly/src/parsers/ast/mod.rs

```diff
@@ -0,0 +1,563 @@
+            let parsed_nodes = self.parse_body_token(tokens, program)?;
+            end_of_nodes = parsed_nodes.is_empty();
+            for node in parsed_nodes {
+                program.body.push(node);
+                empty = false;
+            }
+        }
+
+        // make sure at least one block has been read
+        if empty {
+            let start_op = tokens.read_at(start_pos).expect("no start token");
+            return Err(AssemblyError::empty_block(start_op));
+        }
+
+        // consume the 'end' token
+        match tokens.read() {
+            None => Err(AssemblyError::unmatched_begin(
+                tokens.read_at(program_start).expect("no begin token"),
+            )),
+            Some(token) => match token.parts()[0] {
+                Token::END => token.validate_end(),
+                Token::ELSE => Err(AssemblyError::dangling_else(token)),
+                _ => Err(AssemblyError::unmatched_begin(
+                    tokens.read_at(program_start).expect("no begin token"),
+                )),
+            },
+        }?;
+        tokens.advance();
+
+        Ok(())
+    }
+
+    // Parse procedure from token stream.
+    fn parse_procedure(
+        &mut self,
+        tokens: &mut TokenStream,
+        program: &mut types::Program,
+        allow_export: bool,
+    ) -> Result<types::Procedure, AssemblyError> {
+        let proc_start = tokens.pos();
+
+        // read procedure name and consume the procedure header token
+        let header = tokens.read().expect("missing procedure header");
+        let (label, num_locals, is_export) = header.parse_proc()?;
+        if !allow_export && is_export {
+            return Err(AssemblyError::proc_export_not_allowed(header, &label));
+        }
+        tokens.advance();
+
+        // parse procedure body
+        let body = self.parse_body_token(tokens, program)?;
+
+        // consume the 'end' token
+        match tokens.read() {
+            None => Err(AssemblyError::unmatched_proc(
+                tokens.read_at(proc_start).expect("no proc token"),
+            )),
+            Some(token) => match token.parts()[0] {
+                Token::END => token.validate_end(),
+                _ => Err(AssemblyError::unmatched_proc(
+                    tokens.read_at(proc_start).expect("no proc token"),
+                )),
+            },
+        }?;
+        tokens.advance();
+
+        // build and return the procedure
+        Ok(types::Procedure {
+            name: label,
+            num_locals,
+            is_export,
+            body,
+            index: 0,
+        })
+    }
+
+    // Parses a token from the token stream in a body, which generates a series of AST nodes.
+    fn parse_body_token(
+        &mut self,
+        tokens: &mut TokenStream,
+        program: &mut types::Program,
+    ) -> Result<Vec<Node>, AssemblyError> {
+        let mut nodes = Vec::<Node>::new();
+        if let Some(token) = tokens.read() {
```

Comment:
I think this would be easer to read if we had a separate function for parsing each statement. For example: `parse_if_else`, `parse_while`, parse_repeat` etc.

Reply:
fixed